### PR TITLE
Prioritize HTTP requests over Heal

### LINKED ISSUE #6196: Prioritize HTTP requests over heal operation
**Not using the template on purpose**

With https://github.com/minio/minio/pull/6091 we added low priority
scheduling for usage check routine, such that
other go-routines blocked will be automatically
unblocked and prioritized before usage. 

Based on discussion with @abperiasamy we should take a similar approach
for healing operation where incoming requests to the server are 
prioritised over internal ongoing healing operations

### LINKED ISSUE #6241: Improving Heal performance by allowing parallel heal
When performing Heal on large OBS with a huge amount of small files it can take an extremely long time to complete. 

In our case with have a 60TB OBS with 1MB objects which translate to 60 Million objects. heal manage to process around 10GiB an hour (the setup is with 36 disks). this means that the healing process can take several weeks which isn't reasonable.
This is due to the fact that Minio is processing one object at a time.

in order to allow better performance and give the admin more flexibility, a flag should be added to "mc" which will allow parallelism of the heal operation. Ideally, the parallelism can be changed while the process is running.

This will allow the admin to decide when to give more resources to the "heal" process without affecting the ongoing operation of the cluster 

The way to reproduce the issue is building Minio cluster with a large number of small objects in one bucket and run recursive heal on the entire bucket.

## Your Environment

* Version used (`minio version`): 
Version: 2018-07-23T18:34:49Z
Release-Tag: RELEASE.2018-07-23T18-34-49Z
Commit-ID: 5acc2a6db1fae9923139dfdfab6faa5de48deb73
* Server type and version: 
6 servers with 6 disk of 4TB is on each server
* Operating System and version (`uname -a`): 
3.10.0-693.21.1.el7.x86_64 #1 SMP Wed Mar 7 19:03:37 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux

### PR DESCRIPTION
**Pull request by harshavardhana**
## Description
Prioritize HTTP requests over Heal

## Motivation and Context
Additionally also heal 256 objects at any given
time in parallel.

Fixes #6196
Fixes #6241

## Regression
No

## How Has This Been Tested?
- Create an erasure coded setup with 4 or 6 disks locally using `minio server /tmp/{1...6}`
- Upload a whole bunch of objects using `mc mirror` to this new instance. 
- Delete one or more disks as needed for testing healing `rm -rf /tmp/{1,2}`
- Observe the results of `mc admin heal -r myminio` v/s the current release timing 
  on how fast the healing is. 

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added unit tests to cover my changes.
- [ ] I have added/updated functional tests in [mint](https://github.com/minio/mint). (If yes, add `mint` PR # here: )
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-tls.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :x: [more...](#7401fe14-bbb6-4654-8968-ee9186a3c664) |
| mint-dist-xl.sh | :x: [more...](#eb727b2c-4c94-4cb3-bc47-f6f13be6b833) |
<a name="eb727b2c-4c94-4cb3-bc47-f6f13be6b833"></a>
### 6468-5a13003/mint-dist-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      minikube:32356
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp f8c7913ab825:/mint/log /tmp/mint-logs'
(1/13) Running awscli tests ... done in 1 minutes and 0 seconds
(2/13) Running aws-sdk-go tests ... done in 0 seconds
(3/13) Running aws-sdk-java tests ... done in 3 seconds
(4/13) Running aws-sdk-php tests ... done in 43 seconds
(5/13) Running aws-sdk-ruby tests ... done in 9 seconds
(6/13) Running mc tests ... done in 27 seconds
(7/13) Running minio-dotnet tests ... FAILED in 2 minutes and 17 seconds

Unhandled Exception: System.AggregateException: One or more errors occurred. (One or more errors occurred. (Minio API responded with message=We encountered an internal error, please try again.)) ---> System.AggregateException: One or more errors occurred. (Minio API responded with message=We encountered an internal error, please try again.) ---> Minio.Exceptions.MinioException: Minio API responded with message=We encountered an internal error, please try again.
   at Minio.MinioClient.ParseError(IRestResponse response) in /q/.q/sources/minio-dotnet/Minio/MinioClient.cs:line 362
   at Minio.MinioClient.HandleIfErrorResponse(IRestResponse response, IEnumerable`1 handlers, DateTime startTime) in /q/.q/sources/minio-dotnet/Minio/MinioClient.cs:line 504
   at Minio.MinioClient.ExecuteTaskAsync(IEnumerable`1 errorHandlers, IRestRequest request, CancellationToken cancellationToken) in /q/.q/sources/minio-dotnet/Minio/MinioClient.cs:line 350
   at Minio.MinioClient.PutObjectAsync(String bucketName, String objectName, String uploadId, Int32 partNumber, Byte[] data, Dictionary`2 metaData, CancellationToken cancellationToken) in /q/.q/sources/minio-dotnet/Minio/ApiEndpoints/ObjectOperations.cs:line 483
   at Minio.MinioClient.PutObjectAsync(String bucketName, String objectName, Stream data, Int64 size, String contentType, Dictionary`2 metaData, CancellationToken cancellationToken) in /q/.q/sources/minio-dotnet/Minio/ApiEndpoints/ObjectOperations.cs:line 229
   at Minio.Functional.Tests.FunctionalTest.PutObject_Task(MinioClient minio, String bucketName, String objectName, String fileName, String contentType, Int64 size, Dictionary`2 metaData, MemoryStream mstream) in /mint/run/core/minio-dotnet/FunctionalTest.cs:line 740
   --- End of inner exception stack trace ---
   at System.Threading.Tasks.Task.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at System.Threading.Tasks.Task.Wait()
   at Minio.Functional.Tests.FunctionalTest.RemoveObjects_Test2(MinioClient minio) in /mint/run/core/minio-dotnet/FunctionalTest.cs:line 1725
   --- End of inner exception stack trace ---
   at System.Threading.Tasks.Task.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at System.Threading.Tasks.Task.Wait()
   at Minio.Functional.Tests.FunctionalTest.Main(String[] args) in /mint/run/core/minio-dotnet/FunctionalTest.cs:line 218

Executed 6 out of 13 tests successfully.
```
<a name="7401fe14-bbb6-4654-8968-ee9186a3c664"></a>
### 6468-5a13003/mint-gateway-azure.sh.log:
```
Running with
SERVER_ENDPOINT:      minikube:31835
ACCESS_KEY:           minioazure
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 99126cf0247a:/mint/log /tmp/mint-logs'
(1/13) Running awscli tests ... done in 2 minutes and 43 seconds
(2/13) Running aws-sdk-go tests ... done in 1 seconds
(3/13) Running aws-sdk-java tests ... done in 3 seconds
(4/13) Running aws-sdk-php tests ... done in 47 seconds
(5/13) Running aws-sdk-ruby tests ... done in 36 seconds
(6/13) Running mc tests ... done in 7 minutes and 23 seconds
(7/13) Running minio-dotnet tests ... FAILED in 5 minutes and 1 seconds

Unhandled Exception: System.AggregateException: One or more errors occurred. (Assert.AreEqual failed. Expected:<0x8D619EB0722730A-1>. Actual:<0x8D619EB06A77FF8-1>. ) ---> Microsoft.VisualStudio.TestTools.UnitTesting.AssertFailedException: Assert.AreEqual failed. Expected:<0x8D619EB0722730A-1>. Actual:<0x8D619EB06A77FF8-1>. 
   at Microsoft.VisualStudio.TestTools.UnitTesting.Assert.HandleFail(String assertionName, String message, Object[] parameters)
   at Microsoft.VisualStudio.TestTools.UnitTesting.Assert.AreEqual[T](T expected, T actual, String message, Object[] parameters)
   at Minio.Functional.Tests.FunctionalTest.CopyObject_Test3(MinioClient minio) in /mint/run/core/minio-dotnet/FunctionalTest.cs:line 960
   --- End of inner exception stack trace ---
   at System.Threading.Tasks.Task.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at System.Threading.Tasks.Task.Wait()
   at Minio.Functional.Tests.FunctionalTest.Main(String[] args) in /mint/run/core/minio-dotnet/FunctionalTest.cs:line 223

Executed 6 out of 13 tests successfully.
```

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-tls.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :x: [more...](#2e85be8b-ffdc-4319-8a26-dfd32d8187ad) |
<a name="2e85be8b-ffdc-4319-8a26-dfd32d8187ad"></a>
### 6468-2da3ff2/mint-gateway-azure.sh.log:
```
Running with
SERVER_ENDPOINT:      minikube:31127
ACCESS_KEY:           minioazure
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 2273bb354957:/mint/log /tmp/mint-logs'
(1/13) Running awscli tests ... done in 2 minutes and 51 seconds
(2/13) Running aws-sdk-go tests ... done in 2 seconds
(3/13) Running aws-sdk-java tests ... done in 4 seconds
(4/13) Running aws-sdk-php tests ... done in 50 seconds
(5/13) Running aws-sdk-ruby tests ... done in 40 seconds
(6/13) Running mc tests ... done in 7 minutes and 24 seconds
(7/13) Running minio-dotnet tests ... FAILED in 5 minutes and 22 seconds
{
  "name": "minio-dotnet: CopyObject_Test5",
  "function": "Task<CopyObjectResult> CopyObjectAsync(string bucketName, string objectName, string destBucketName, string destObjectName = null, CopyConditions copyConditions = null, CancellationToken cancellationToken = default(CancellationToken))",
  "args": {
    "bucketName": "miniodotnetb562pmerqnfs9gr",
    "objectName": "miniodotnetf7dcivv046",
    "destBucketName": "miniodotnet9halbxi64e407ek",
    "destObjectName": "miniodotnett2t3e8ew3d",
    "data": "6MB",
    "size": "6MB"
  },
  "duration": 8289,
  "status": "FAIL",
  "alert": "",
  "message": "Minio API responded with message=A header you provided implies functionality that is not implemented",
  "error": "A header you provided implies functionality that is not implemented: Minio.Exceptions.MinioException: Minio API responded with message=A header you provided implies functionality that is not implemented\n   at Minio.MinioClient.ParseError(IRestResponse response) in /q/.q/sources/minio-dotnet/Minio/MinioClient.cs:line 362\n   at Minio.MinioClient.HandleIfErrorResponse(IRestResponse response, IEnumerable`1 handlers, DateTime startTime) in /q/.q/sources/minio-dotnet/Minio/MinioClient.cs:line 504\n   at Minio.MinioClient.ExecuteTaskAsync(IEnumerable`1 errorHandlers, IRestRequest request, CancellationToken cancellationToken) in /q/.q/sources/minio-dotnet/Minio/MinioClient.cs:line 350\n   at Minio.MinioClient.CopyObjectRequestAsync(String bucketName, String objectName, String destBucketName, String destObjectName, CopyConditions copyConditions, Dictionary`2 customHeaders, String resource, CancellationToken cancellationToken, Type type) in /q/.q/sources/minio-dotnet/Minio/ApiEndpoints/ObjectOperations.cs:line 986\n   at Minio.MinioClient.MultipartCopyUploadAsync(String bucketName, String objectName, String destBucketName, String destObjectName, CopyConditions copyConditions, Int64 copySize, Dictionary`2 metadata, CancellationToken cancellationToken) in /q/.q/sources/minio-dotnet/Minio/ApiEndpoints/ObjectOperations.cs:line 1045\n   at Minio.MinioClient.CopyObjectAsync(String bucketName, String objectName, String destBucketName, String destObjectName, CopyConditions copyConditions, Dictionary`2 metadata, CancellationToken cancellationToken) in /q/.q/sources/minio-dotnet/Minio/ApiEndpoints/ObjectOperations.cs:line 939\n   at Minio.Functional.Tests.FunctionalTest.CopyObject_Test5(MinioClient minio) in /mint/run/core/minio-dotnet/FunctionalTest.cs:line 1059"
}

Executed 6 out of 13 tests successfully.
```

File: cmd/admin-heal-ops.go [hunk 1]
```diff
@@ -625,14 +626,35 @@ func (h *healSequence) healBucket(bucket string) error {
...
+			for globalHTTPServer.GetRequestCount() > 0 && waitCount > 0 {
+				waitCount--
+				time.Sleep(1 * time.Second)
+			}
+		}
+
+		// Heal 256 objects at a time.
 		objectInfos, err := objectAPI.ListObjectsHeal(h.ctx, bucket,
-			h.objPrefix, marker, "", 1000)
+			h.objPrefix, marker, "", 256)
```
Comment (donatello) [inline cmd/admin-heal-ops.go]: Any reason for 256 here? It looks arbitrary and seems high for small number of node deployments.

At least 1 cpu is needed to actively heal an object and trying to heal many simultaneously is not going to be faster, as most will simply wait for cpu/disk. It seems like `numCPU * numNodes` is a better degree of parallelism. WIth a large number of healing threads, normal processing will be slowed down (after the 1 minute delay is done).

**Review by donatello — CHANGES_REQUESTED** [review]

File: cmd/admin-heal-ops.go (hunk 1 above)
Comment (harshavardhana) [reply cmd/admin-heal-ops.go]: 256 is simply an arbitrary number. Perhaps we can do that runtime.CPU() * len(globalEndpoints)

**Review by donatello — APPROVED** [review]

**Review by poornas — APPROVED** [review]
LGTM

**Pull request merged**